File: Python/ByTier/E/1774_E_Two_Chess_Pieces.py
# Problem: CF 1774 E - Two Chess Pieces
# https://codeforces.com/contest/1774/problem/E

"""
Algorithm: Tree Traversal and Dynamic Programming on Trees
Approach:
1. Build a tree from input edges.
2. For each node, track:
   - Whether it's part of the path for piece 1 or 2 (seq1, seq2)
   - Maximum distance from root for piece 1 or 2 (dist1, dist2)
3. Perform DFS to calculate minimum steps needed to ensure that both pieces never exceed distance d from each other.

Time Complexity: O(n)
Space Complexity: O(n)
"""

import sys
from array import array
from collections import deque

input = lambda: sys.stdin.readline().rstrip("\r\n")
inp = lambda dtype: [dtype(x) for x in input().split()]
debug = lambda *x: print(*x, file=sys.stderr)
ceil1 = lambda a, b: (a + b - 1) // b
Mint, Mlong = 2**31 - 1, 2**63 - 1
out, tests = [], 1


class graph:
    def __init__(self, n):
        self.n = n
        # Initialize adjacency list
        self.gdict = [array("i") for _ in range(n + 1)]

    def add_edge(self, node1, node2):
        # Add bidirectional edges
        self.gdict[node1].append(node2)
        self.gdict[node2].append(node1)

    def subtree(self, root):
        # BFS to get parent info and child nodes 
        queue, visit = deque([root]), array("b", [0] * (self.n + 1))
        child = array("i")
        # Initialize parent array
        self.par, visit[root] = array("i", [-1] * (self.n + 1)), True

        while queue:
            s = queue.popleft()

            for i1 in self.gdict[s]:
                if not visit[i1]:
                    queue.append(i1)
                    visit[i1] = True
                    child.append(i1)
                    self.par[i1] = s

        # Traverse nodes from leaves to root to compute distances
        for ch in child[::-1]:
            # Update maximum distances for piece 1 and 2
            dist1[self.par[ch]] = max(dist1[self.par[ch]], dist1[ch] + seq2[ch])
            dist2[self.par[ch]] = max(dist2[self.par[ch]], dist2[ch] + seq1[ch])
            # Update presence of required nodes
            seq1[self.par[ch]] |= seq1[ch]
            seq2[self.par[ch]] |= seq2[ch]

    def dfs(self, root):
        # DFS to count necessary moves
        stk, vis = array("i", [root]), array("b", [0] * (self.n + 1))
        vis[root], ret = 1, 0

        while stk:
            par = stk.pop()
            for ch in self.gdict[par]:
                if not vis[ch]:
                    vis[ch] = 1
                    stk.append(ch)

                    # If path is required or distance constraint violated for piece 1
                    if seq1[ch] or dist1[ch] + 1 > d:
                        ret += 2
                    # If path is required or distance constraint violated for piece 2
                    if seq2[ch] or dist2[ch] + 1 > d:
                        ret += 2

        return ret


for _ in range(tests):
    n, d = inp(int)
    g = graph(n)
    # Read edges and build graph
    for _ in range(n - 1):
        u, v = inp(int)
        g.add_edge(u, v)

    # Initialize arrays for tracking paths and distances
    seq1, seq2 = array("b", [0] * (n + 1)), array("b", [0] * (n + 1))
    dist1, dist2 = array("i", [0] * (n + 1)), array("i", [0] * (n + 1))
    
    # Mark required nodes for first piece
    for i in array("i", inp(int))[1:]:
        seq1[i] = 1
    # Mark required nodes for second piece
    for i in array("i", inp(int))[1:]:
        seq2[i] = 1

    # Compute subtree info
    g.subtree(1)
    # Compute minimum steps required
    out.append(g.dfs(1))

print("\n".join(map(str, out)))


# https://github.com/VaHiX/CodeForces/

File: CPP/ByTier/E/1774_E_Two_Chess_Pieces.cpp
// Problem: CF 1774 E - Two Chess Pieces
// https://codeforces.com/contest/1774/problem/E

/*
 * Problem: E. Two Chess Pieces
 * 
 * Purpose: Given a tree with n nodes and two pieces starting at node 1, determine the minimum steps
 *          to visit all required nodes by both pieces while maintaining a maximum distance d between them.
 * 
 * Algorithm:
 * - Build adjacency list representation of the tree.
 * - Track which nodes each piece must visit using a boolean matrix.
 * - Use DFS to traverse the tree and compute the coverage of each piece.
 * - For each node, if either piece needs to visit it, count it toward the total.
 * - Return 2 * total visited nodes as each node must be entered and exited.
 * 
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 * 
 * Techniques:
 * - Tree traversal using DFS
 * - Bitmask tracking of node visits
 * - Distance constraint enforcement via tracking
 */

#include <algorithm>
#include <iostream>
#include <vector>

#define all(x) begin(x), end(x)
#define rall(x) rbegin(x), rend(x)
using namespace std;
using ll = long long;

void solve() {
  int n, d;
  cin >> n >> d;
  vector<vector<int>> g(n);
  for (int i = 0; i < n - 1; i++) {
    int u, v;
    cin >> u >> v;
    u--;
    v--;
    g[u].push_back(v);
    g[v].push_back(u);
  }
  vector<vector<int>> vis(2, vector<int>(n));
  for (int i = 0; i < 2; ++i) {
    int m;
    cin >> m;
    for (int j = 0; j < m; ++j) {
      int v;
      cin >> v;
      v--;
      vis[i][v] = 1;
    }
  }
  vector<int> path(n);
  vis[0][0] = vis[1][0] = 1;
  int len = 0;
  auto dfs = [&](int par, int curr, auto &&dfs) -> void {
    // If we've moved beyond distance d, mark the old node's influence on current node
    if (len >= d) {
      vis[0][path[len - d]] |= vis[1][curr];
      vis[1][path[len - d]] |= vis[0][curr];
    }
    path[len++] = curr;
    for (auto nei : g[curr]) {
      if (nei != par)
        dfs(curr, nei, dfs);
    }
    if (par != -1) {
      // Propagate visit information up the tree
      vis[0][par] |= vis[0][curr];
      vis[1][par] |= vis[1][curr];
    }
    len--;
  };
  dfs(-1, 0, dfs);
  int sol = 0;
  for (int i = 1; i < n; ++i) {
    sol += vis[0][i];
    sol += vis[1][i];
  }
  cout << 2 * sol << endl;
}

int main() {
  ios_base::sync_with_stdio(false);
  cin.tie(0);
  int t = 1;
  while (t--)
    solve();
#ifdef LOCAL_TIME
  cerr << "Time elapsed: " << 1.0 * clock() / CLOCKS_PER_SEC << "s\n";
#endif
}


// https://github.com/VaHiX/CodeForces/